File: fixtures/scale/typescript/large/src/ciphers/Carousel.ts
// Rotation transposition cipher (carousel).
export class Carousel {
  static offset(text: string, key: number): number {
    return (key + 5) % Math.max(1, text.length);
  }

  static encrypt(text: string, key: number): string {
    const n = Carousel.offset(text, key);
    return text.slice(n) + text.slice(0, n);
  }

  static decrypt(text: string, key: number): string {
    const n = Carousel.offset(text, key);
    return text.slice(text.length - n) + text.slice(0, text.length - n);
  }
}

File: fixtures/scale/lua/large/ciphers/carousel.lua
local function carousel_offset(text, key)
  return (key + 5) % math.max(1, #text)
end

local function carousel_encrypt(text, key)
  local n = carousel_offset(text, key)
  return text:sub(n + 1) .. text:sub(1, n)
end

local function carousel_decrypt(text, key)
  local n = carousel_offset(text, key)
  return text:sub(#text - n + 1) .. text:sub(1, #text - n)
end

return { carousel_offset = carousel_offset, carousel_encrypt = carousel_encrypt, carousel_decrypt = carousel_decrypt }
